Spurious list expectation for x-deprecated property

### Expected Behavior

Successful API generation from schema

### Actual Behaviour

Generation fails with `class java.lang.String cannot be cast to class java.util.List`

The problem goes away when this property (line 84 of test-3.1.json) is changed from

```json
"x-deprecated" : "replaced with ClientOpEnum"
```

to 

```json
"x-deprecated" : [ "replaced with ClientOpEnum" ]
```


### Steps To Reproduce

https://github.com/veselov/micronaut-test/tree/issue1

Run Maven build

Observe OpenAPI generation failure.

```
Caused by: java.lang.ClassCastException: class java.lang.String cannot be cast to class java.util.List (java.lang.String and java.util.List are in module java.base of loader 'bootstrap')
    at io.micronaut.openapi.generator.AbstractMicronautJavaCodegen.updateEnumVarsWithExtensions (AbstractMicronautJavaCodegen.java:2308)
    at org.openapitools.codegen.DefaultCodegen.updateCodegenPropertyEnum (DefaultCodegen.java:7012)
    at org.openapitools.codegen.DefaultCodegen.postProcessModelsEnum (DefaultCodegen.java:863)
    at org.openapitools.codegen.languages.AbstractJavaCodegen.postProcessModels (AbstractJavaCodegen.java:2020)
    at org.openapitools.codegen.DefaultGenerator.processModels (DefaultGenerator.java:1849)
    at org.openapitools.codegen.DefaultGenerator.generateModels (DefaultGenerator.java:546)
    at org.openapitools.codegen.DefaultGenerator.generateModels (DefaultGenerator.java:476)
    at org.openapitools.codegen.DefaultGenerator.generate (DefaultGenerator.java:1325)
    at io.micronaut.openapi.generator.MicronautCodeGeneratorEntryPoint.generate (MicronautCodeGeneratorEntryPoint.java:125)
    at io.micronaut.maven.openapi.AbstractOpenApiMojo.execute (AbstractOpenApiMojo.java:492)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
```



### Environment Information

_No response_

### Example Application

https://github.com/veselov/micronaut-test/tree/issue1

### Version

4.10.0

--- a/openapi-generator/src/main/java/io/micronaut/openapi/generator/AbstractMicronautJavaCodegen.java
+++ b/openapi-generator/src/main/java/io/micronaut/openapi/generator/AbstractMicronautJavaCodegen.java
@@ -101,9 +101,11 @@ import static io.micronaut.openapi.generator.Utils.NULL_STRING;
 import static io.micronaut.openapi.generator.Utils.addEnumParamsForConverters;
 import static io.micronaut.openapi.generator.Utils.addStrValueToEnum;
 import static io.micronaut.openapi.generator.Utils.calcQueryValueFormat;
+import static io.micronaut.openapi.generator.Utils.findEnumVar;
 import static io.micronaut.openapi.generator.Utils.isDateType;
 import static io.micronaut.openapi.generator.Utils.normalizeExtraAnnotations;
 import static io.micronaut.openapi.generator.Utils.processDuplicateVars;
+import static io.micronaut.openapi.generator.Utils.processEnumExt;
 import static io.micronaut.openapi.generator.Utils.processGenericAnnotations;
 import static io.micronaut.openapi.generator.Utils.readListOfStringsProperty;
 import static io.swagger.v3.parser.util.SchemaTypeUtil.BYTE_FORMAT;
@@ -1629,6 +1631,15 @@ public abstract class AbstractMicronautJavaCodegen<T extends GeneratorOptionsBui
                 processGenericAnnotations(op.returnProperty, useBeanValidation, isGenerateHardNullable(), false, false, false, false, false, false);
                 op.returnType = op.returnProperty.vendorExtensions.get("typeWithEnumWithGenericAnnotations").toString();
             }
+
+            var deprecatedMessage = op.vendorExtensions.get("x-deprecated-message");
+            var xDeprecated = op.vendorExtensions.get("x-deprecated");
+            if (deprecatedMessage == null && xDeprecated instanceof String xDeprecatedStr) {
+                deprecatedMessage = xDeprecatedStr;
+            }
+            if (deprecatedMessage != null) {
+                op.vendorExtensions.put("x-deprecated-message", deprecatedMessage.toString());
+            }
         }
 
         additionalProperties.put("enumImports", enumImports);
@@ -2247,6 +2258,16 @@ public abstract class AbstractMicronautJavaCodegen<T extends GeneratorOptionsBui
                 }
             }
             model.allVars = allVars;
+
+            var deprecatedMessage = model.vendorExtensions.get("x-deprecated-message");
+            var xDeprecated = model.vendorExtensions.get("x-deprecated");
+            if (deprecatedMessage == null && xDeprecated instanceof String xDeprecatedStr) {
+                deprecatedMessage = xDeprecatedStr;
+            }
+            if (deprecatedMessage != null) {
+                model.vendorExtensions.put("x-deprecated-message", deprecatedMessage.toString());
+            }
+
             model.vendorExtensions.put("requiredVars", requiredVars);
             model.vendorExtensions.put("optionalVars", optionalVars);
             model.vendorExtensions.put("areRequiredVarsAndReadOnlyVars", !requiredVarsWithoutDiscriminator.isEmpty() && !model.readOnlyVars.isEmpty());
@@ -2300,52 +2321,47 @@ public abstract class AbstractMicronautJavaCodegen<T extends GeneratorOptionsBui
 
     @Override
     protected void updateEnumVarsWithExtensions(List<Map<String, Object>> enumVars, Map<String, Object> vendorExtensions, String dataType) {
-        super.updateEnumVarsWithExtensions(enumVars, vendorExtensions, dataType);
         if (vendorExtensions == null) {
             return;
         }
 
-        var xDeprecated = (List<Object>) vendorExtensions.get("x-deprecated");
+        processEnumExt(enumVars, vendorExtensions, "enumDescription", List.of("x-enum-descriptions", "x-xl4-enum-doc"));
+        processEnumExt(enumVars, vendorExtensions, "enumDeprecatedMessage", List.of("x-enum-deprecated-messages", "x-xl4-enum-deprecated"));
+
+        var deprecatedExt = vendorExtensions.get("x-deprecated");
+        List<?> xDeprecated = null;
+        Map<?, ?> xDeprecatedMap = null;
+        if (deprecatedExt instanceof List<?> deprecatedList) {
+            xDeprecated = deprecatedList;
+        } else if (deprecatedExt instanceof Map<?, ?> deprecatedMap) {
+            xDeprecated = new ArrayList<>(deprecatedMap.keySet());
+            xDeprecatedMap = deprecatedMap;
+        }
+        var deprecatedMessagesExt = vendorExtensions.get("x-enum-deprecated-messages");
+        if (deprecatedMessagesExt == null) {
+            deprecatedMessagesExt = vendorExtensions.get("x-xl4-enum-deprecated");
+        }
+        if (xDeprecated == null && deprecatedMessagesExt instanceof Map<?, ?> deprecatedMap) {
+            xDeprecated = new ArrayList<>(deprecatedMap.keySet());
+            xDeprecatedMap = deprecatedMap;
+        }
         if (xDeprecated != null && !xDeprecated.isEmpty()) {
             for (var deprecatedItem : xDeprecated) {
-                Map<String, Object> foundEnumVar = null;
-                for (var enumVar : enumVars) {
-                    var isString = (boolean) enumVar.get("isString");
-                    var value = (String) enumVar.get("value");
-                    if (!isString) {
-                        if (value.startsWith("(short)")) {
-                            value = value.replace("(short) ", "");
-                        } else if (value.startsWith("(byte)")) {
-                            value = value.replace("(byte) ", "");
-                        }
-                        var argPos = value.indexOf('(');
-                        // case for BigDecimal
-                        if (argPos >= 0) {
-                            value = value.substring(argPos + 1, value.indexOf(')'));
-                        }
-                        var upperValue = value.toUpperCase(Locale.ENGLISH);
-                        if (upperValue.endsWith("F")
-                            || upperValue.endsWith("L")
-                            || upperValue.endsWith("D")) {
-                            value = value.substring(0, value.length() - 1);
-                        }
-                        if (!value.contains("'")) {
-                            value = value.replace("'", "");
-                        }
-                        if (!value.contains("\"")) {
-                            value = "\"" + value + "\"";
-                        }
-                    }
-                    if (value.equals("\"" + deprecatedItem + '"')) {
-                        foundEnumVar = enumVar;
-                        break;
-                    }
+                Map<String, Object> foundEnumVar = findEnumVar(deprecatedItem.toString(), enumVars);
+                if (foundEnumVar == null) {
+                    continue;
                 }
-                if (foundEnumVar != null) {
-                    foundEnumVar.put("deprecated", true);
+                foundEnumVar.put("deprecated", true);
+                if (xDeprecatedMap != null) {
+                    var deprecatedMessage = xDeprecatedMap.get(deprecatedItem);
+                    if (deprecatedMessage instanceof String deprecatedMessageStr) {
+                        foundEnumVar.put("enumDeprecatedMessage", deprecatedMessageStr);
+                    }
                 }
             }
         }
+
+        processEnumExt(enumVars, vendorExtensions, "name", List.of("x-enum-varnames"));
 
         var baseType = (String) vendorExtensions.get("baseType");
         for (var enumVar : enumVars) {
@@ -2465,6 +2481,15 @@ public abstract class AbstractMicronautJavaCodegen<T extends GeneratorOptionsBui
     }
 
     private void processProperty(CodegenProperty property, boolean isServer, CodegenModel model, Map<String, ModelsMap> models) {
+
+        var deprecatedMessage = property.vendorExtensions.get("x-deprecated-message");
+        var xDeprecated = property.vendorExtensions.get("x-deprecated");
+        if (deprecatedMessage == null && xDeprecated instanceof String xDeprecatedStr) {
+            deprecatedMessage = xDeprecatedStr;
+        }
+        if (deprecatedMessage != null) {
+            property.vendorExtensions.put("x-deprecated-message", deprecatedMessage.toString());
+        }
 
         property.vendorExtensions.put("inRequiredArgsConstructor", !property.isReadOnly || isServer);
         property.vendorExtensions.put("isServer", isServer);

--- a/openapi-generator/src/main/java/io/micronaut/openapi/generator/AbstractMicronautKotlinCodegen.java
+++ b/openapi-generator/src/main/java/io/micronaut/openapi/generator/AbstractMicronautKotlinCodegen.java
@@ -111,11 +111,13 @@ import static io.micronaut.openapi.generator.Utils.NULL_STRING;
 import static io.micronaut.openapi.generator.Utils.addEnumParamsForConverters;
 import static io.micronaut.openapi.generator.Utils.addStrValueToEnum;
 import static io.micronaut.openapi.generator.Utils.calcQueryValueFormat;
+import static io.micronaut.openapi.generator.Utils.findEnumVar;
 import static io.micronaut.openapi.generator.Utils.isDateType;
 import static io.micronaut.openapi.generator.Utils.normalizeExtraAnnotations;
 import static io.micronaut.openapi.generator.Utils.normalizeStr;
 import static io.micronaut.openapi.generator.Utils.normalizeStrValue;
 import static io.micronaut.openapi.generator.Utils.processDuplicateVars;
+import static io.micronaut.openapi.generator.Utils.processEnumExt;
 import static io.micronaut.openapi.generator.Utils.processGenericAnnotations;
 import static io.micronaut.openapi.generator.Utils.readListOfStringsProperty;
 import static io.swagger.v3.parser.util.SchemaTypeUtil.BYTE_FORMAT;
@@ -1319,8 +1321,14 @@ public abstract class AbstractMicronautKotlinCodegen<T extends GeneratorOptionsB
 
     private void normalizeTextsInOperation(CodegenOperation op) {
 
-        if (op.vendorExtensions.containsKey("x-deprecated-message")) {
-            op.vendorExtensions.put("x-deprecated-message-normalized", normalizeStr(op.vendorExtensions.get("x-deprecated-message").toString()));
+        var deprecatedMessage = op.vendorExtensions.get("x-deprecated-message");
+        var xDeprecated = op.vendorExtensions.get("x-deprecated");
+        if (deprecatedMessage == null && xDeprecated instanceof String xDeprecatedStr) {
+            deprecatedMessage = xDeprecatedStr;
+            op.vendorExtensions.put("x-deprecated-message", xDeprecatedStr);
+        }
+        if (deprecatedMessage != null) {
+            op.vendorExtensions.put("x-deprecated-message-normalized", normalizeStr(deprecatedMessage.toString()));
         }
 
         if (op.vendorExtensions.containsKey("x-roles")) {
@@ -2357,8 +2365,14 @@ public abstract class AbstractMicronautKotlinCodegen<T extends GeneratorOptionsB
                 model.vendorExtensions.put("requiredParentVarsWithoutDiscriminator", requiredParentVarsWithoutDiscriminator);
             }
 
-            if (model.vendorExtensions.containsKey("x-deprecated-message")) {
-                model.vendorExtensions.put("x-deprecated-message-normalized", normalizeStr((String) model.vendorExtensions.get("x-deprecated-message")));
+            var deprecatedMessage = model.vendorExtensions.get("x-deprecated-message");
+            var xDeprecated = model.vendorExtensions.get("x-deprecated");
+            if (deprecatedMessage == null && xDeprecated instanceof String xDeprecatedStr) {
+                deprecatedMessage = xDeprecatedStr;
+                model.vendorExtensions.put("x-deprecated-message", xDeprecatedStr);
+            }
+            if (deprecatedMessage != null) {
+                model.vendorExtensions.put("x-deprecated-message-normalized", normalizeStr(deprecatedMessage.toString()));
             }
             model.vendorExtensions.put("descriptionNormalized", normalizeStr(model.description));
             model.vendorExtensions.put("requiredVarsWithoutDiscriminator", requiredVarsWithoutDiscriminator);
@@ -2482,49 +2496,48 @@ public abstract class AbstractMicronautKotlinCodegen<T extends GeneratorOptionsB
 
     @Override
     protected void updateEnumVarsWithExtensions(List<Map<String, Object>> enumVars, Map<String, Object> vendorExtensions, String dataType) {
-        super.updateEnumVarsWithExtensions(enumVars, vendorExtensions, dataType);
         if (vendorExtensions == null) {
             return;
         }
-        var xDeprecated = (List<Object>) vendorExtensions.get("x-deprecated");
+
+        processEnumExt(enumVars, vendorExtensions, "enumDescription", List.of("x-enum-descriptions", "x-xl4-enum-doc"));
+        processEnumExt(enumVars, vendorExtensions, "enumDeprecatedMessage", List.of("x-enum-deprecated-messages", "x-xl4-enum-deprecated"), true);
+
+        var deprecatedExt = vendorExtensions.get("x-deprecated");
+        List<?> xDeprecated = null;
+        Map<?, ?> xDeprecatedMap = null;
+        if (deprecatedExt instanceof List<?> deprecatedList) {
+            xDeprecated = deprecatedList;
+        } else if (deprecatedExt instanceof Map<?, ?> deprecatedMap) {
+            xDeprecated = new ArrayList<>(deprecatedMap.keySet());
+            xDeprecatedMap = deprecatedMap;
+        }
+        var deprecatedMessagesExt = vendorExtensions.get("x-enum-deprecated-messages");
+        if (deprecatedMessagesExt == null) {
+            deprecatedMessagesExt = vendorExtensions.get("x-xl4-enum-deprecated");
+        }
+        if (xDeprecated == null && deprecatedMessagesExt instanceof Map<?, ?> deprecatedMap) {
+            xDeprecated = new ArrayList<>(deprecatedMap.keySet());
+            xDeprecatedMap = deprecatedMap;
+        }
         if (xDeprecated != null && !xDeprecated.isEmpty()) {
             for (var deprecatedItem : xDeprecated) {
-                Map<String, Object> foundEnumVar = null;
-                for (var enumVar : enumVars) {
-                    var isString = (boolean) enumVar.get("isString");
-                    var value = (String) enumVar.get("value");
-                    if (!isString) {
-                        if (value.startsWith("(short)")) {
-                            value = value.replace("(short) ", "");
-                        }
-                        var argPos = value.indexOf('(');
-                        // case for BigDecimal
-                        if (argPos >= 0) {
-                            value = value.substring(argPos + 1, value.indexOf(')'));
-                        }
-                        var upperValue = value.toUpperCase(Locale.ENGLISH);
-                        if (upperValue.endsWith("F")
-                            || upperValue.endsWith("L")
-                            || upperValue.endsWith("D")) {
-                            value = value.substring(0, value.length() - 1);
-                        }
-                        if (!value.contains("'")) {
-                            value = value.replace("'", "");
-                        }
-                        if (!value.contains("\"")) {
-                            value = "\"" + value + "\"";
-                        }
-                    }
-                    if (value.equals("\"" + deprecatedItem + '"')) {
-                        foundEnumVar = enumVar;
-                        break;
-                    }
+                Map<String, Object> foundEnumVar = findEnumVar(deprecatedItem.toString(), enumVars);
+                if (foundEnumVar == null) {
+                    continue;
                 }
-                if (foundEnumVar != null) {
-                    foundEnumVar.put("deprecated", true);
+                foundEnumVar.put("deprecated", true);
+                if (xDeprecatedMap != null) {
+                    var deprecatedMessage = xDeprecatedMap.get(deprecatedItem);
+                    if (deprecatedMessage instanceof String deprecatedMessageStr) {
+                        foundEnumVar.put("enumDeprecatedMessage", deprecatedMessageStr);
+                        foundEnumVar.put("enumDeprecatedMessageNormalized", normalizeStr(deprecatedMessageStr));
+                    }
                 }
             }
         }
+
+        processEnumExt(enumVars, vendorExtensions, "name", List.of("x-enum-varnames"));
 
         var baseType = (String) vendorExtensions.get("baseType");
         for (var enumVar : enumVars) {
@@ -2656,8 +2669,14 @@ public abstract class AbstractMicronautKotlinCodegen<T extends GeneratorOptionsB
         if (property.description != null) {
             property.vendorExtensions.put("descriptionNormalized", normalizeStr(property.description));
         }
-        if (property.vendorExtensions.containsKey("x-deprecated-message")) {
-            property.vendorExtensions.put("x-deprecated-message-normalized", normalizeStr((String) property.vendorExtensions.get("x-deprecated-message")));
+        var deprecatedMessage = property.vendorExtensions.get("x-deprecated-message");
+        var xDeprecated = property.vendorExtensions.get("x-deprecated");
+        if (deprecatedMessage == null && xDeprecated instanceof String xDeprecatedStr) {
+            deprecatedMessage = xDeprecatedStr;
+            property.vendorExtensions.put("x-deprecated-message", xDeprecatedStr);
+        }
+        if (deprecatedMessage != null) {
+            property.vendorExtensions.put("x-deprecated-message-normalized", normalizeStr(deprecatedMessage.toString()));
         }
         normalizeStrValue("x-pattern-message", property.vendorExtensions);
         normalizeStrValue("x-size-message", property.vendorExtensions);

--- a/openapi-generator/src/main/java/io/micronaut/openapi/generator/Utils.java
+++ b/openapi-generator/src/main/java/io/micronaut/openapi/generator/Utils.java
@@ -783,4 +783,73 @@ public final class Utils {
 
         return true;
     }
+
+    public static void processEnumExt(List<Map<String, Object>> enumVars, Map<String, Object> vendorExtensions, String key, List<String> extensionKeys) {
+        processEnumExt(enumVars, vendorExtensions, key, extensionKeys, false);
+    }
+
+    public static void processEnumExt(List<Map<String, Object>> enumVars, Map<String, Object> vendorExtensions, String key, List<String> extensionKeys, boolean withNormalization) {
+        Object value = null;
+        for (var extKey : extensionKeys) {
+            value = vendorExtensions.get(extKey);
+            if (value != null) {
+                break;
+            }
+        }
+        if (value instanceof List<?> valuesList) {
+            int size = Math.min(enumVars.size(), valuesList.size());
+            for (int i = 0; i < size; i++) {
+                enumVars.get(i).put(key, valuesList.get(i));
+                if (withNormalization) {
+                    enumVars.get(i).put(key + "Normalized", normalizeStr(valuesList.get(i).toString()));
+                }
+            }
+        } else if (value instanceof Map<?, ?> valuesMap) {
+            for (var entry : valuesMap.entrySet()) {
+                Map<String, Object> foundEnumVar = findEnumVar(entry.getKey().toString(), enumVars);
+                if (foundEnumVar == null) {
+                    continue;
+                }
+                foundEnumVar.put(key, entry.getValue().toString());
+                if (withNormalization) {
+                    foundEnumVar.put(key + "Normalized", normalizeStr(entry.getValue().toString()));
+                }
+            }
+        }
+    }
+
+    public static Map<String, Object> findEnumVar(String enumConstName, List<Map<String, Object>> enumVars) {
+        for (var enumVar : enumVars) {
+            var isString = (boolean) enumVar.get("isString");
+            var value = (String) enumVar.get("value");
+            if (!isString) {
+                if (value.startsWith("(short)")) {
+                    value = value.replace("(short) ", "");
+                } else if (value.startsWith("(byte)")) {
+                    value = value.replace("(byte) ", "");
+                }
+                var argPos = value.indexOf('(');
+                // case for BigDecimal
+                if (argPos >= 0) {
+                    value = value.substring(argPos + 1, value.indexOf(')'));
+                }
+                var upperValue = value.toUpperCase(Locale.ENGLISH);
+                if (upperValue.endsWith("F")
+                    || upperValue.endsWith("L")
+                    || upperValue.endsWith("D")) {
+                    value = value.substring(0, value.length() - 1);
+                }
+                if (!value.contains("'")) {
+                    value = value.replace("'", "");
+                }
+                if (!value.contains("\"")) {
+                    value = "\"" + value + "\"";
+                }
+            }
+            if (value.equals("\"" + enumConstName + '"')) {
+                return enumVar;
+            }
+        }
+        return null;
+    }
 }

--- a/openapi-generator/src/main/resources/templates/java-micronaut/common/model/enum.mustache
+++ b/openapi-generator/src/main/resources/templates/java-micronaut/common/model/enum.mustache
@@ -31,12 +31,21 @@
 {{#formatSingleLine}}public enum {{>common/model/enumName}}{{#vendorExtensions.x-implements}}{{#-first}} implements {{{.}}}{{/-first}}{{^-first}}, {{{.}}}{{/-first}}{{/vendorExtensions.x-implements}}{{/formatSingleLine}} {
 
     {{#allowableValues}}
-        {{#enumVars}}
+        {{#enumVars}}{{#formatNoEmptyLines}}
             {{#enumDescription}}
     /**
      * {{{enumDescription}}}
+{{#deprecated}}
+     *
+     * @deprecated{{#enumDeprecatedMessage}} {{{.}}}{{/enumDeprecatedMessage}}{{/deprecated}}
      */
             {{/enumDescription}}
+            {{^enumDescription}}{{#deprecated}}
+    /**
+     *
+     * @deprecated{{#enumDeprecatedMessage}} {{{.}}}{{/enumDeprecatedMessage}}
+     */
+            {{/deprecated}}{{/enumDescription}}
             {{#withXml}}
     @XmlEnumValue({{#isInteger}}"{{/isInteger}}{{#isDouble}}"{{/isDouble}}{{#isLong}}"{{/isLong}}{{#isFloat}}"{{/isFloat}}{{{value}}}{{#isInteger}}"{{/isInteger}}{{#isDouble}}"{{/isDouble}}{{#isLong}}"{{/isLong}}{{#isFloat}}"{{/isFloat}})
             {{/withXml}}
@@ -46,6 +55,7 @@
     @JsonProperty({{{strValue}}})
     {{{name}}}({{{value}}}),{{#-last}}
     ;{{/-last}}
+        {{/formatNoEmptyLines}}
         {{/enumVars}}
     {{/allowableValues}}
 

--- a/openapi-generator/src/main/resources/templates/kotlin-micronaut/common/model/enum.mustache
+++ b/openapi-generator/src/main/resources/templates/kotlin-micronaut/common/model/enum.mustache
@@ -4,7 +4,7 @@
  * @param value The value represented by this enum
 {{#isDeprecated}}
  *
- * @deprecated {{#vendorExtensions.x-deprecated-message}} {{{.}}}{{/vendorExtensions.x-deprecated-message}}
+ * @deprecated{{#vendorExtensions.x-deprecated-message}} {{{.}}}{{/vendorExtensions.x-deprecated-message}}
 {{/isDeprecated}}
  */
 {{#isDeprecated}}
@@ -32,21 +32,31 @@
 ){{#vendorExtensions.x-implements}}{{#-first}} : {{{.}}}{{/-first}}{{^-first}}, {{{.}}}{{/-first}}{{/vendorExtensions.x-implements}} {
 
     {{#allowableValues}}
-        {{#enumVars}}
+        {{#enumVars}}{{#formatNoEmptyLines}}
             {{#enumDescription}}
     /**
-     * {{enumDescription}}
+     * {{{enumDescription}}}
+{{#deprecated}}
+     *
+     * @deprecated{{#enumDeprecatedMessage}} {{{.}}}{{/enumDeprecatedMessage}}{{/deprecated}}
      */
             {{/enumDescription}}
+            {{^enumDescription}}{{#deprecated}}
+    /**
+     *
+     * @deprecated{{#enumDeprecatedMessage}} {{{.}}}{{/enumDeprecatedMessage}}
+     */
+            {{/deprecated}}{{/enumDescription}}
             {{#withXml}}
     @XmlEnumValue({{#isInteger}}"{{/isInteger}}{{#isDouble}}"{{/isDouble}}{{#isLong}}"{{/isLong}}{{#isFloat}}"{{/isFloat}}{{{value}}}{{#isInteger}}"{{/isInteger}}{{#isDouble}}"{{/isDouble}}{{#isLong}}"{{/isLong}}{{#isFloat}}"{{/isFloat}})
             {{/withXml}}
             {{#deprecated}}
-    @Deprecated("")
+    @Deprecated("{{#enumDeprecatedMessageNormalized}}{{{.}}}{{/enumDeprecatedMessageNormalized}}")
             {{/deprecated}}
     @JsonProperty({{{strValueNormalized}}})
     {{{name}}}({{{valueNormalized}}}),{{#-last}}
     ;{{/-last}}
+        {{/formatNoEmptyLines}}
         {{/enumVars}}
     {{/allowableValues}}
 

--- a/openapi-generator/src/test/java/io/micronaut/openapi/generator/JavaMicronautClientCodegenTest.java
+++ b/openapi-generator/src/test/java/io/micronaut/openapi/generator/JavaMicronautClientCodegenTest.java
@@ -2205,15 +2205,156 @@ class JavaMicronautClientCodegenTest extends AbstractMicronautCodegenTest {
 
         String path = outputPathApi + "src/main/java/org/openapitools/";
         assertFileContains(path + "config/EnumConverterClientConfig.java", """
-                    @Bean
-                    public TypeConverter<String, ParamEnum> toEnumParamEnum() {
-                        return (v, c, ctx) -> Optional.of(ParamEnum.fromValue(v));
-                    }
+                @Bean
+                public TypeConverter<String, ParamEnum> toEnumParamEnum() {
+                    return (v, c, ctx) -> Optional.of(ParamEnum.fromValue(v));
+                }
+            
+                @Bean
+                public TypeConverter<ParamEnum, String> toStrParamEnum() {
+                    return (v, c, ctx) -> Optional.of(v.getValue());
+                }
+            """);
+    }
+
+    @Test
+    void testEnumExtensions() {
+
+        var codegen = new JavaMicronautClientCodegen();
+        String outputPathApi = generateFiles(codegen, "src/test/resources/3_0/enum-extensions.yml");
+
+        String path = outputPathApi + "src/main/java/org/openapitools/";
+        assertFileContains(path + "model/EnumWithList.java", """
+                 *
+                 * @deprecated This is enum deprecated message
+                 */
+                @Deprecated
+                """,
+            """
+                    /**
+                     * doc1
+                     */
+                    @JsonProperty("CGI")
+                    A("CGI"),
                 
-                    @Bean
-                    public TypeConverter<ParamEnum, String> toStrParamEnum() {
-                        return (v, c, ctx) -> Optional.of(v.getValue());
-                    }
+                    /**
+                     *
+                     * @deprecated This is deprecated message1
+                     */
+                    @Deprecated
+                    @JsonProperty("CGU")
+                    B("CGU"),
+                
+                    /**
+                     * doc3
+                     *
+                     * @deprecated This is deprecated message2
+                     */
+                    @Deprecated
+                    @JsonProperty("BRB")
+                    C("BRB"),
+                
+                    @JsonProperty("NUM")
+                    NUM("NUM"),
+                """);
+        assertFileContains(path + "model/EnumWithMap.java", """
+                 *
+                 * @deprecated This is enum deprecated message
+                 */
+                @Deprecated
+                """,
+            """
+                    /**
+                     * doc1
+                     */
+                    @JsonProperty("CGI")
+                    A("CGI"),
+                
+                    /**
+                     *
+                     * @deprecated This is deprecated message1
+                     */
+                    @Deprecated
+                    @JsonProperty("CGU")
+                    B("CGU"),
+                
+                    /**
+                     * doc3
+                     *
+                     * @deprecated This is deprecated message2
+                     */
+                    @Deprecated
+                    @JsonProperty("BRB")
+                    C("BRB"),
+                
+                    @JsonProperty("NUM")
+                    NUM("NUM"),
+                """);
+        assertFileContains(path + "model/EnumWithMap2.java", """
+                 *
+                 * @deprecated This is enum deprecated message
+                 */
+                @Deprecated
+                """,
+            """
+                    /**
+                     * doc1
+                     */
+                    @JsonProperty("CGI")
+                    A("CGI"),
+                
+                    /**
+                     *
+                     * @deprecated This is deprecated message1
+                     */
+                    @Deprecated
+                    @JsonProperty("CGU")
+                    B("CGU"),
+                
+                    /**
+                     * doc3
+                     *
+                     * @deprecated This is deprecated message2
+                     */
+                    @Deprecated
+                    @JsonProperty("BRB")
+                    C("BRB"),
+                
+                    @JsonProperty("NUM")
+                    NUM("NUM"),
+                """);
+        assertFileContains(path + "model/EnumWithDifferentExts.java", """
+                 *
+                 * @deprecated This is enum deprecated message
+                 */
+                @Deprecated
+                """,
+            """
+                    /**
+                     * doc1
+                     */
+                    @JsonProperty("CGI")
+                    A("CGI"),
+                
+                    /**
+                     *
+                     * @deprecated This is deprecated message1
+                     */
+                    @Deprecated
+                    @JsonProperty("CGU")
+                    B("CGU"),
+                
+                    /**
+                     * doc3
+                     *
+                     * @deprecated This is deprecated message2
+                     */
+                    @Deprecated
+                    @JsonProperty("BRB")
+                    C("BRB"),
+                
+                    @JsonProperty("NUM")
+                    NUM("NUM"),
                 """);
     }
 }

--- a/openapi-generator/src/test/java/io/micronaut/openapi/generator/KotlinMicronautClientCodegenTest.java
+++ b/openapi-generator/src/test/java/io/micronaut/openapi/generator/KotlinMicronautClientCodegenTest.java
@@ -2587,4 +2587,145 @@ class KotlinMicronautClientCodegenTest extends AbstractMicronautCodegenTest {
                 fun toStrParamEnum() = TypeConverter<ParamEnum, String> { v, _, _ -> Optional.of(v.value) }
             """);
     }
+
+    @Test
+    void testEnumExtensions() {
+
+        var codegen = new KotlinMicronautClientCodegen();
+        String outputPathApi = generateFiles(codegen, "src/test/resources/3_0/enum-extensions.yml");
+
+        String path = outputPathApi + "src/main/kotlin/org/openapitools/";
+        assertFileContains(path + "model/EnumWithList.kt", """
+                 *
+                 * @deprecated This is enum deprecated message
+                 */
+                @Deprecated("This is enum deprecated message")
+                """,
+            """
+                    /**
+                     * doc1
+                     */
+                    @JsonProperty("CGI")
+                    A("CGI"),
+                
+                    /**
+                     *
+                     * @deprecated This is deprecated message1
+                     */
+                    @Deprecated("This is deprecated message1")
+                    @JsonProperty("CGU")
+                    B("CGU"),
+                
+                    /**
+                     * doc3
+                     *
+                     * @deprecated This is deprecated message2
+                     */
+                    @Deprecated("This is deprecated message2")
+                    @JsonProperty("BRB")
+                    C("BRB"),
+                
+                    @JsonProperty("NUM")
+                    NUM("NUM"),
+                """);
+        assertFileContains(path + "model/EnumWithMap.kt", """
+                 *
+                 * @deprecated This is enum deprecated message
+                 */
+                @Deprecated("This is enum deprecated message")
+                """,
+            """
+                    /**
+                     * doc1
+                     */
+                    @JsonProperty("CGI")
+                    A("CGI"),
+                
+                    /**
+                     *
+                     * @deprecated This is deprecated message1
+                     */
+                    @Deprecated("This is deprecated message1")
+                    @JsonProperty("CGU")
+                    B("CGU"),
+                
+                    /**
+                     * doc3
+                     *
+                     * @deprecated This is deprecated message2
+                     */
+                    @Deprecated("This is deprecated message2")
+                    @JsonProperty("BRB")
+                    C("BRB"),
+                
+                    @JsonProperty("NUM")
+                    NUM("NUM"),
+                """);
+        assertFileContains(path + "model/EnumWithMap2.kt", """
+                 *
+                 * @deprecated This is enum deprecated message
+                 */
+                @Deprecated("This is enum deprecated message")
+                """,
+            """
+                    /**
+                     * doc1
+                     */
+                    @JsonProperty("CGI")
+                    A("CGI"),
+                
+                    /**
+                     *
+                     * @deprecated This is deprecated message1
+                     */
+                    @Deprecated("This is deprecated message1")
+                    @JsonProperty("CGU")
+                    B("CGU"),
+                
+                    /**
+                     * doc3
+                     *
+                     * @deprecated This is deprecated message2
+                     */
+                    @Deprecated("This is deprecated message2")
+                    @JsonProperty("BRB")
+                    C("BRB"),
+                
+                    @JsonProperty("NUM")
+                    NUM("NUM"),
+                """);
+        assertFileContains(path + "model/EnumWithDifferentExts.kt", """
+                 *
+                 * @deprecated This is enum deprecated message
+                 */
+                @Deprecated("This is enum deprecated message")
+                """,
+            """
+                    /**
+                     * doc1
+                     */
+                    @JsonProperty("CGI")
+                    A("CGI"),
+                
+                    /**
+                     *
+                     * @deprecated This is deprecated message1
+                     */
+                    @Deprecated("This is deprecated message1")
+                    @JsonProperty("CGU")
+                    B("CGU"),
+                
+                    /**
+                     * doc3
+                     *
+                     * @deprecated This is deprecated message2
+                     */
+                    @Deprecated("This is deprecated message2")
+                    @JsonProperty("BRB")
+                    C("BRB"),
+                
+                    @JsonProperty("NUM")
+                    NUM("NUM"),
+                """);
+    }
 }

--- a/openapi-generator/src/test/java/io/micronaut/openapi/generator/KotlinMicronautServerCodegenTest.java
+++ b/openapi-generator/src/test/java/io/micronaut/openapi/generator/KotlinMicronautServerCodegenTest.java
@@ -1571,10 +1571,20 @@ class KotlinMicronautServerCodegenTest extends AbstractMicronautCodegenTest {
                 
                 }
                 """
-            );
+        );
 
         assertFileContains(path + "model/DollarGetDollarStringsDollarQueryVarParameter.kt",
             """
+                /**
+                 * Gets or Sets $get$Strings_$queryVar_parameter
+                 *
+                 * @param value The value represented by this enum
+                 *
+                 * @deprecated Deprecated message with $dollarSign
+                 */
+                @Deprecated("Deprecated message with \\$dollarSign")
+                @Serdeable
+                @Generated("io.micronaut.openapi.generator.KotlinMicronautServerCodegen")
                 enum class DollarGetDollarStringsDollarQueryVarParameter(
                     @get:JsonValue
                     val value: String,
@@ -1585,10 +1595,13 @@ class KotlinMicronautServerCodegenTest extends AbstractMicronautCodegenTest {
                      */
                     @JsonProperty("ref1/\\$ref")
                     REF1__REF("ref1/\\$ref"),
+                
                     /**
                      * desc enumConst2 with $dollarSign
+                     *
+                     * @deprecated Deprecated message with $dollarSign2
                      */
-                    @Deprecated("")
+                    @Deprecated("Deprecated message with \\$dollarSign2")
                     @JsonProperty("ref2/\\$ref")
                     REF2__REF("ref2/\\$ref"),
                     ;

--- a/openapi-generator/src/test/resources/3_0/dollar-sign.yml
+++ b/openapi-generator/src/test/resources/3_0/dollar-sign.yml
@@ -72,6 +72,8 @@ paths:
           schema:
             type: string
             default: ref2/$ref
+            deprecated: true
+            x-deprecated-message: Deprecated message with $dollarSign
             enum:
               - ref1/$ref
               - ref2/$ref
@@ -80,6 +82,9 @@ paths:
               - desc enumConst2 with $dollarSign
             x-deprecated:
               - ref2/$ref
+            x-enum-deprecated-messages:
+              - ""
+              - Deprecated message with $dollarSign2
         - name: $pathVar
           in: path
           description: desc pathVar with $dollarSign

--- a/openapi-generator/src/test/resources/3_0/enum-extensions.yml
+++ b/openapi-generator/src/test/resources/3_0/enum-extensions.yml
@@ -1,0 +1,90 @@
+openapi: 3.1.0
+info:
+  title: title
+  description: description
+  version: ''
+servers:
+  - url: /
+components:
+  schemas:
+    EnumWithList:
+      type: string
+      deprecated: true
+      x-deprecated-message: This is enum deprecated message
+      enum:
+        - CGI
+        - CGU
+        - BRB
+        - NUM
+      x-deprecated:
+        - CGU
+        - BRB
+      x-enum-varnames:
+        - A
+        - B
+        - C
+      x-enum-descriptions:
+        - doc1
+        -
+        - doc3
+      x-enum-deprecated-messages:
+        - ""
+        - This is deprecated message1
+        - This is deprecated message2
+    EnumWithMap:
+      type: string
+      deprecated: true
+      x-deprecated-message: This is enum deprecated message
+      enum:
+        - CGI
+        - CGU
+        - BRB
+        - NUM
+      x-deprecated:
+        CGU: This is deprecated message1
+        BRB: This is deprecated message2
+      x-enum-varnames:
+        CGI: A
+        CGU: B
+        BRB: C
+      x-enum-descriptions:
+        CGI: doc1
+        BRB: doc3
+    EnumWithMap2:
+      type: string
+      deprecated: true
+      x-deprecated-message: This is enum deprecated message
+      enum:
+        - CGI
+        - CGU
+        - BRB
+        - NUM
+      x-enum-deprecated-messages:
+        CGU: This is deprecated message1
+        BRB: This is deprecated message2
+      x-enum-varnames:
+        CGI: A
+        CGU: B
+        BRB: C
+      x-enum-descriptions:
+        CGI: doc1
+        BRB: doc3
+    EnumWithDifferentExts:
+      type: string
+      deprecated: true
+      x-deprecated: This is enum deprecated message
+      enum:
+        - CGI
+        - CGU
+        - BRB
+        - NUM
+      x-xl4-enum-deprecated:
+        CGU: This is deprecated message1
+        BRB: This is deprecated message2
+      x-enum-varnames:
+        CGI: A
+        CGU: B
+        BRB: C
+      x-xl4-enum-doc:
+        CGI: doc1
+        BRB: doc3

